### ISSUE TITLE
Mock -ModuleName scoping issues

### ISSUE THREAD

**Issue by dlwyatt** [pester/Pester#204]
`Mock -ModuleName` is causing the `MockWith` script block to be executed inside of the module, instead of out in the test code's scope which is where every other parameter filter and mock script block is executed.  For consistency, mocks should be executed in the test script scope; `-ModuleName` should just inject the mock into the module to intercept calls there.  The current state of affairs can cause problems with resolving variables out in the script scope.  For example:

``` posh
Describe 'Mocking without a module' {
    function TestFunction { param ([string] $Parameter) }

    $returnValue = 'The Return Value'
    $paramValue = 'The Parameter'

    Mock TestFunction { return $returnValue } -ParameterFilter { $Parameter -eq $paramValue }

    # This works
    It 'Resolves variables from within the test script' {
        TestFunction -Parameter $paramValue | Should Be $returnValue
    }
}

Describe 'Mocking in a module' {
    New-Module -Name TestModule {
        function PublicFunction { InternalFunction -Parameter 'The Parameter' }
        function InternalFunction {
            param ([string] $Parameter)
        }

        Export-ModuleMember -Function PublicFunction
    } | Import-Module -Force

    $returnValue = 'The Return Value'
    $paramValue = 'The Parameter'

    Mock InternalFunction { return $returnValue } -ModuleName TestModule  -ParameterFilter { $Parameter -eq $paramValue }

    # This fails
    It 'Resolves variables from within the test script' {
        TestModule\PublicFunction | Should Be $returnValue
    }

    It 'Called the mock' {
        Assert-MockCalled InternalFunction -ModuleName TestModule
    }

    Remove-Module TestModule -Force
}
```

This code demonstrates that even though the mock was called in the second Describe, it returned the wrong value.  This is because the MockWith block was executed in a scope where $returnValue doesn't exist.  Notice that the ParameterFilter blocks are executed out in the test script's scope in both cases, allowing the $paramValue variable to be used successfully.  It's weird to have the parameter filter and MockWith blocks executed in different scopes.

Fixing this would be a breaking change from v3.0's behavior, so I haven't done it yet.  We should probably do so in v4, though.

**Comment by vors** [comment]
During work on [ZLocation bad scenario test](https://github.com/vors/ZLocation/blob/master/ZLocation.Tests/ZLocation.Tests.ps1#L42) I had a lot of frustration trying to mock function in **nested modules** (ZLocation.Storage.psm1 and ZLocation.Search.psm1 used to be NestedModules (declared in `.psd1`)). Should I use `-ModuleName` as a nestedModule or root module?

I suspect that nested modules have a separate scoping model in PowerShell and I'm not sure that  pester respect it.

It was a late night and I just replaced NestedModules by direct `Import-Module` to make it mockable.

I just want to check, have anybody look at this scenario, maybe it can be coordinated with other Mock work.

I would provide a small repro in a separate issue.

**Comment by dlwyatt** [comment]
I don't think I've tried to work with nested modules and Pester yet.  I can look into it, though.

**Comment by it-praktyk** [comment]
@dlwyatt, can it be closed now?

**Comment by dlwyatt** [comment]
I'll double check later today and open a PR if needed.

**Comment by dlwyatt** [comment]
Nevermind.  If I change this, it breaks several other things.  There are workarounds if anyone actually runs into this situation:

- Just use `InModuleScope` for this sort of thing.
- Call `GetNewClosure()` on the script block that's being passed to `-MockWith`
- Use global variables (blech).

**Comment by it-praktyk** [comment]
> Call GetNewClosure() on the script block that's being passed to -MockWith 

GetNewClosure() is undocumented - I've reopened issue than to remember about this.

**Comment by chadattapfin** [comment]
I ran into this issue as well. Thank you @dlwyatt for the temporary workaround. Instead of using the psd1 scope, I used the inner module scope and am off and running again.

**Comment by nohwnd** [comment]
Fixed in #1281 bodies and parameter filters now run in the test scope.
